### PR TITLE
GH-43044: [R] So-called non-API entry points

### PR DESCRIPTION
**Pull request by jonkeane**
### Rationale for this change

CRAN 

### What changes are included in this PR?

Remove so-called non-api calls

### Are these changes tested?

With tests

### Are there any user-facing changes?

Hopefully not

**This PR contains a "Critical Fix".**
* GitHub Issue: #43044

### CODE REVIEW

File: r/src/arrow_cpp11.h [hunk 1]
```diff
@@ -44,9 +44,6 @@
 #define UTF8_MASK (1 << 3)
 #define ASCII_MASK (1 << 6)
```
Comment (nealrichardson) [inline r/src/arrow_cpp11.h:45]: I think you can delete these too, they were only used by the macros you removed

**Review by nealrichardson — CHANGES_REQUESTED** [review]
I think this is ok, just one suggestion for something else to remove. AFAICT we only call `r_string_size` on a `STRING_ELT` from a `STRSXP`, and similarly for `utf8_string`/`utf8_strings`. `Rf_translateCharUTF8` should always work on those inputs.

Here's the R internals version of IS_ASCII etc.; `LEVELS` is defined as `sxpinfo.gp`, an opaque bitmask. Even if this change turns out not to be correct, I don't think we want to be touching `sxpinfo.gp` for anything ourselves. https://github.com/wch/r-source/blob/0ebd0b0b574847597ac19fbf0bbbf31e3197e84b/src/include/Defn.h#L832-L853

**Comment by nealrichardson** [comment]
> I think this is ok, just one suggestion for something else to remove. AFAICT we only call `r_string_size` on a `STRING_ELT` from a `STRSXP`, and similarly for `utf8_string`/`utf8_strings`. `Rf_translateCharUTF8` should always work on those inputs.
> 
> Here's the R internals version of IS_ASCII etc.; `LEVELS` is defined as `sxpinfo.gp`, an opaque bitmask. Even if this change turns out not to be correct, I don't think we want to be touching `sxpinfo.gp` for anything ourselves. https://github.com/wch/r-source/blob/0ebd0b0b574847597ac19fbf0bbbf31e3197e84b/src/include/Defn.h#L832-L853

Ah great, you've found the line in the scriptures to confirm that https://github.com/apache/arrow/issues/43044#issuecomment-2212292694

File: r/src/arrow_cpp11.h (hunk 1 above)
Comment (jonkeane) [reply r/src/arrow_cpp11.h]: Ah right of course

**Comment by jonkeane** [comment]
> Ah great, you've found the line in the scriptures to confirm that https://github.com/apache/arrow/issues/43044#issuecomment-2212292694

It took many nights of studying the holy text

**Pull request merged**